src/compiler/core.ts

```diff
@@ -1312,4 +1333,49 @@ namespace ts {
             : ((fileName) => fileName.toLowerCase());
     }
 
+    /**
+     * This isn't the strictest deep equal, but it's good enough for us
+     *  - +0 === -0 (though who really wants to consider them different?)
+     *  - arguments and arrays can be equal (both typeof === object, both have enumerable keys)
+     *  - doesn't inspect es6 iterables (not that they're used in this code base)
+     *  - doesn't inspect regex toString value (so only references to the same regex are equal)
+     *  - doesn't inspect date primitive number value (so only references to the same date are equal)
+     */
+    export function deepEqual(a: any, b: any, memo?: [any, any][]): boolean {

```

Comment:
Extension arguments objects can be literally any object, and that's what's getting compared. I think the only thing this handles which I _shouldn't_ (provided good input from any language service wrappers or other API consumers) need to worry about for that is circularly referential objects. (Which is why all the shortcomings listed in the comment were OK)